pkg/datapath/linux/arp/ping.go

```diff
@@ -0,0 +1,202 @@
+// Copyright 2018-2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package arp
+
+import (
+	"bytes"
+	"errors"
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
+	"net"
+	"time"
+)
+
+var (
+	ErrNotImplemented = errors.New("not implemented")
+	ErrL2Unreachable  = errors.New("interface can't reach the IP address")
+
+	timeout = 1 * time.Second
+)
+
+var defaultSerializeOpts = gopacket.SerializeOptions{
+	FixLengths:       true,
+	ComputeChecksums: true,
+}
+
+// PingOverIface performs arping request to the 'dst' IP address over the 'iface' and
+// returns the hardware address (MAC) of the destination if it is reachable
+func PingOverIface(iface net.Interface, dst net.IP) (net.HardwareAddr, error) {
+	src, err := findSourceIP(iface, dst)
+	if err != nil {
+		return nil, err
+	}
+
+	p, err := newPinger(&iface, src)
+	if err != nil {
+		return nil, err
+	}
+	defer p.close()
+
+	if err := p.setDeadline(time.Now().Add(timeout)); err != nil {
+		return nil, err
+	}
+
+	return p.resolve(dst)
+}
+
+var _ net.Addr = &Addr{}
+
+type Addr struct {
+	net.HardwareAddr
+}
+
+func (a *Addr) Network() string {
+	return "raw"
+}
+
+type pinger struct {
+	c     net.PacketConn
+	ip    net.IP
+	iface *net.Interface
+}
+
+func (p *pinger) close() {
+	_ = p.c.Close()
+}
+
+func (p *pinger) setDeadline(t time.Time) error {
+	return p.c.SetDeadline(t)
+}
+
+func (p *pinger) resolve(ip net.IP) (net.HardwareAddr, error) {
+	if err := p.request(ip); err != nil {
+		return nil, err
+	}
+
+	for {
+		resp, err := p.read()
+		if err != nil {
+			return nil, err
+		}
+
+		if resp.Operation != layers.ARPReply || !bytes.Equal(resp.SourceProtAddress, []byte(ip.To4())) {
```

Comment:
Its not obvious to me how this error happens. No problems with the code, just would like to understand why we might receive a '!= layers.ARPeply' or maybe its just for completeness?

Reply:
It was mostly for completeness and to show the intention. I won't mind removing it if you think it is not needed.